RecoBTag/FeatureTools/plugins/DeepBoostedJetTagInfoProducer.cc

```diff
@@ -245,122 +217,159 @@ void DeepBoostedJetTagInfoProducer::produce(edm::Event& iEvent, const edm::Event
+          << "Cannot convert to either pat::PackedCandidate or "
+             "reco::PFCandidate";
     }
     puppi_wgt_cache[cand.key()] = wgt;
     return wgt;
   };
 
   std::vector<reco::CandidatePtr> daughters;
-  for (const auto& cand : jet.daughterPtrVector()){
+  for (const auto &cand : jet.daughterPtrVector()) {
     // remove particles w/ extremely low puppi weights
-    if ((puppiWgt(cand)) < 0.01) continue;
+    if ((puppiWgt(cand)) < 0.01)
+      continue;
     daughters.push_back(cand);
   }
   // sort by (Puppi-weighted) pt
   if (!has_puppi_weighted_daughters_) {
-    std::sort(daughters.begin(), daughters.end(), [&puppi_wgt_cache](const reco::CandidatePtr& a, const reco::CandidatePtr& b){
-      return puppi_wgt_cache.at(a.key())*a->pt() > puppi_wgt_cache.at(b.key())*b->pt(); });
-  }else{
-    std::sort(daughters.begin(), daughters.end(), [](const reco::CandidatePtr& a, const reco::CandidatePtr& b){ return a->pt() > b->pt(); });
+    std::sort(daughters.begin(), daughters.end(),
+              [&puppi_wgt_cache](const reco::CandidatePtr &a,
+                                 const reco::CandidatePtr &b) {
+                return puppi_wgt_cache.at(a.key()) * a->pt() >
+                       puppi_wgt_cache.at(b.key()) * b->pt();
+              });
+  } else {
+    std::sort(daughters.begin(), daughters.end(),
+              [](const reco::CandidatePtr &a, const reco::CandidatePtr &b) {
+                return a->pt() > b->pt();
+              });
   }
 
   // reserve space
-  for (const auto &name : particle_features_) { fts.reserve(name, daughters.size()); }
+  for (const auto &name : particle_features_) {
+    fts.reserve(name, daughters.size());
+  }
 
-  auto useTrackProperties = [&](const reco::PFCandidate* reco_cand) {
-    const auto* trk = reco_cand->bestTrack();
-    return trk!=nullptr && trk->pt()>min_pt_for_track_properties_;
+  auto useTrackProperties = [&](const reco::PFCandidate *reco_cand) {
+    const auto *trk = reco_cand->bestTrack();
+    return trk != nullptr && trk->pt() > min_pt_for_track_properties_;
   };
 
-  for (const auto& cand : daughters){
-    const auto* packed_cand = dynamic_cast<const pat::PackedCandidate*>(&(*cand));
-    const auto* reco_cand = dynamic_cast<const reco::PFCandidate*>(&(*cand));
+  for (const auto &cand : daughters) {
+    const auto *packed_cand =
+        dynamic_cast<const pat::PackedCandidate *>(&(*cand));
+    const auto *reco_cand = dynamic_cast<const reco::PFCandidate *>(&(*cand));
 
     auto puppiP4 = cand->p4();
-    if (packed_cand){
+    if (packed_cand) {
       if (!has_puppi_weighted_daughters_) {
         puppiP4 *= puppi_wgt_cache.at(cand.key());
       }
 
-      fts.fill("pfcand_hcalFrac", packed_cand->hcalFraction());
+      fts.fill("pfcand_hcalFrac", (packed_cand->isIsolatedChargedHadron())
+                                      ? packed_cand->rawHcalFraction()
+                                      : packed_cand->hcalFraction());
```

Comment:
if the goal is to preserve the old behavior which somewhat benefited from the "ID" provided by the isIsolatedChargedHadron to have hcalFraction not zero, then the change above should pass hcalFraction only for HF and neutral hadrons.